### PR TITLE
[Monitor][Query] Support non-public clouds

### PR DESCRIPTION
**Pull request by pvaneck**
This adds support and documents non-public cloud usage for LogsQueryClient and MetricsQueryClient.

There are two URIs we are concerned with: 

- the Log Analytics API endpoint (e.g., `https://api.loganalytics.io/v1`)
- the audience for determining the credential policy scope (`https://api.loganalytics.io`)

The endpoint passed in is expected to be the full endpoint including the `v1`, and the audience is expected to be the base URL of the endpoint. This logic is added to `LogsQueryClient`.

Also adjust tests to be able to accompany different cloud endpoints.
Verified on Azure China Cloud by running live tests there using the weekly pipeline.

Closes: https://github.com/Azure/azure-sdk-for-python/issues/27167

### CODE REVIEW

File: sdk/monitor/azure-monitor-query/README.md [hunk 1]
```diff
@@ -63,6 +63,17 @@ async_logs_client = LogsQueryClient(credential)
 async_metrics_client = MetricsQueryClient(credential)
 ```
 
+#### Configure clients for non-public Azure clouds
+
+By default, `LogsQueryClient` and `MetricsQueryClient` are configured to connect to the public Azure cloud. These can be configured to connect to non-public Azure clouds by passing in the correct `endpoint` argument: For example:
```
Comment (scottaddie) [inline sdk/monitor/azure-monitor-query/README.md:68]:
```suggestion
By default, `LogsQueryClient` and `MetricsQueryClient` are configured to connect to the public Azure cloud. These clients can be configured to connect to non-public Azure clouds via the `endpoint` argument: For example:
```

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_logs_query_client.py [hunk 2]
```diff
@@ -43,20 +44,22 @@ class LogsQueryClient(object): # pylint: disable=client-accepts-api-version-keyw
...
+    :keyword endpoint: The endpoint to connect to. Defaults to 'https://api.loganalytics.io/v1'.
     :paramtype endpoint: Optional[str]
     """
 
     def __init__(self, credential: TokenCredential, **kwargs: Any) -> None:
-        endpoint = kwargs.pop("endpoint", "https://api.loganalytics.io")
+        endpoint = kwargs.pop("endpoint", "https://api.loganalytics.io/v1")
         if not endpoint.startswith("https://") and not endpoint.startswith("http://"):
             endpoint = "https://" + endpoint
+        parsed_endpoint = urlparse(endpoint)
```
Comment (KarishmaGhiya) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_logs_query_client.py:55]: For a user defined endpoint too, we should be appending the version v1. JS customer has reported an issue on this - https://github.com/Azure/azure-sdk-for-js/issues/22408

File: sdk/monitor/azure-monitor-query/azure/monitor/query/aio/_logs_query_client_async.py [hunk 3]
```diff
@@ -34,15 +35,18 @@ class LogsQueryClient(object): # pylint: disable=client-accepts-api-version-keyw
...
             endpoint = "https://" + endpoint
+        parsed_endpoint = urlparse(endpoint)
+        # Assume audience is the base URL of the endpoint, unless a value is explicitly passed in.
+        audience = kwargs.pop("audience", f"{parsed_endpoint.scheme}://{parsed_endpoint.netloc}")
         self._endpoint = endpoint
         auth_policy = kwargs.pop("authentication_policy", None)
         self._client = MonitorQueryClient(
             credential=credential,
-            authentication_policy=auth_policy or get_authentication_policy(credential, endpoint),
-            endpoint=self._endpoint.rstrip('/') + "/v1",
```
Comment (KarishmaGhiya) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/aio/_logs_query_client_async.py:45]: curious, why did you remove the version number? @pvaneck

Comment (pvaneck) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/aio/_logs_query_client_async.py:45]: The API endpoint includes `v1` and is defaulted to `https://api.loganalytics.io/v1` right above. If a user passes in a different endpoint, I'd rather not be appending additional things and just use what we get. So if a user wanted to use the Azure China Log Analytics endpoint, they'd explicitly pass in `endpoint="https://api.loganalytics.azure.cn/v1"`.

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_logs_query_client.py (hunk 2 above)
Comment (pvaneck) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/_logs_query_client.py:55]: I think the guidance for that issue should be that the user pass in `"https://api.loganalytics.io/v1"` as the endpoint, but then the client should strip the v1 and provide that as the audience for the credential scope (i.e., `https://api.loganalytics.io/.default`). Here a user can also explicitly pass in an audience in the off chance that the hostname differs from the endpoint hostname.

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py [hunk 4]
```diff
@@ -16,14 +16,12 @@
```
Comment (kristapratico) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:16]: does _helpers_async.py need this same update?

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_logs_query_client.py [hunk 5]
```diff
@@ -43,20 +44,22 @@ class LogsQueryClient(object): # pylint: disable=client-accepts-api-version-keyw
...
     :param credential: The credential to authenticate the client.
     :type credential: ~azure.core.credentials.TokenCredential
-    :keyword endpoint: The endpoint to connect to. Defaults to 'https://api.loganalytics.io'.
+    :keyword endpoint: The endpoint to connect to. Defaults to 'https://api.loganalytics.io/v1'.
     :paramtype endpoint: Optional[str]
     """
 
     def __init__(self, credential: TokenCredential, **kwargs: Any) -> None:
-        endpoint = kwargs.pop("endpoint", "https://api.loganalytics.io")
+        endpoint = kwargs.pop("endpoint", "https://api.loganalytics.io/v1")
```
Comment (kristapratico) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_logs_query_client.py:52]: Not for this PR, but should we accept an `api_version` keyword only argument in the client?

Comment (pvaneck) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/_logs_query_client.py:52]: Yea, I think this would make sense for MetricsQueryClient since the underlying operations include `api_version` in the requests. No such thing for LogsQueryClient, however, since API versioning is based on what endpoint the user submits a request to.

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py (hunk 4 above)
Comment (pvaneck) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:16]: Yea, it does. Good catch! Completely forgot to update that as well. I am now realizing that `get_authentication_policy` and `get_metrics_authentication_policy` are exactly the same now, so will remove one in each.

**Review by kristapratico — APPROVED** [review]
